array_api_compat/numpy/_typing.py

```diff
@@ -1,46 +1,33 @@
 from __future__ import annotations
 
-__all__ = [
-    "ndarray",
-    "Device",
-    "Dtype",
-]
+__all__ = ["Array", "DType", "Device"]
+_all_ignore = ["np"]
 
-import sys
-from typing import (
-    Literal,
-    Union,
-    TYPE_CHECKING,
-)
+from typing import Literal, Union
 
-from numpy import (
-    ndarray,
-    dtype,
-    int8,
-    int16,
-    int32,
-    int64,
-    uint8,
-    uint16,
-    uint32,
-    uint64,
-    float32,
-    float64,
-)
+import numpy as np
+from numpy import ndarray as Array
 
 Device = Literal["cpu"]
-if TYPE_CHECKING or sys.version_info >= (3, 9):
-    Dtype = dtype[Union[
-        int8,
-        int16,
-        int32,
-        int64,
-        uint8,
-        uint16,
-        uint32,
-        uint64,
-        float32,
-        float64,
-    ]]
-else:
-    Dtype = dtype
+try:
+    DType = np.dtype[
+        Union[
+            np.intp,
+            np.int8,
+            np.int16,
+            np.int32,
+            np.int64,
+            np.uint8,
+            np.uint16,
+            np.uint32,
+            np.uint64,
+            np.float32,
+            np.float64,
+            np.complex64,
+            np.complex128,
+            np.bool_,
+        ]
+    ]
+except TypeError:
+    # NumPy 1.x on Python 3.9 and 3.10
+    DType = np.dtype
```

Comment:
type-checker wouldn't know what to make of a `try/catch` like this:

```suggestion
if TYPE_CHECKING:
    DType: TypeAlias = np.dtype[
        np.bool_ 
        | np.integer 
        | np.single  # same as float32
        | np.double  # slightly broader than float64 in numpy==2.2.*
        | np.csingle  # same as complex64
        | np.cdouble  # slightly broader than complex128 in numpy==2.2.*
    ]
else:
    DType = np.dtype
```

and don't forget the `from typing import TYPE_CHECKING`

Reply:
please no `np.single`, `np.double` etc. Whenever these are used, the usage should drift towards explicit bit-width names, not the other way around.

Reply:
In most cases I agree. But here it's a sneaky workaround for the fact that on `numpy==2.2`, the `float64` is annotated as a proper subtype of `floating[_64Bit]`, while `double` is (still) an alias to `floating[_64Bit]`. 
The consequence is that `x: np.float64 = np.double()` will be rejected, since `double` is a broader type than `float64`. So if you want to accept *both* `double` and `float64`, then you should use `x: double`. Because if you use `x: float64`, then you're excluding `double` and anything else that resolves to `floating[_64Bit]`.